record-and-playback/presentation/scripts/publish/presentation.rb

```diff
@@ -580,251 +545,202 @@ def events_parse_shape(shapes, event, current_presentation, current_slide, times
   shape = {}
   # Common properties
   shape[:in] = timestamp
-  shape[:type] = event.at_xpath('type').text
-  shape[:data_points] = event.at_xpath('dataPoints').text.split(',').map { |p| p.to_f }
+  shape_type = shape[:type] = event.at_xpath('type').text
+  shape_data_points = shape[:data_points] = event.at_xpath('dataPoints').text.split(',').map(&:to_f)
+
   # These can be missing in old BBB versions, there are fallbacks
-  user_id = event.at_xpath('userId')
-  shape[:user_id] = user_id.text if !user_id.nil?
-  shape_id = event.at_xpath('id')
-  shape[:id] = shape_id.text if !shape_id.nil?
-  status = event.at_xpath('status')
-  shape[:status] = status.text if !status.nil?
-  shape[:shape_id] = $svg_shape_id
-  $svg_shape_id += 1
+  user_id = event.at_xpath('userId')&.text
+  shape[:user_id] = user_id if user_id
+
+  shape_id = event.at_xpath('id')&.text
+  shape[:id] = shape_id if shape_id
+
+  status = event.at_xpath('status')&.text
+  shape_status = shape[:status] = status if status
+  draw_id = shape[:shape_id] = @svg_shape_id
+  @svg_shape_id += 1
 
   # Some shape-specific properties
-  if shape[:type] == 'pencil' or shape[:type] == 'rectangle' or
-      shape[:type] == 'ellipse' or shape[:type] == 'triangle' or
-      shape[:type] == 'line'
+  if %w[ellipse line pencil rectangle triangle].include?(shape_type)
     shape[:color] = color_to_hex(event.at_xpath('color').text)
-    thickness = event.at_xpath('thickness')
-    unless thickness
-      BigBlueButton.logger.warn("Draw #{shape[:shape_id]} Shape #{shape[:shape_unique_id]} ID #{shape[:id]} is missing thickness")
+    unless (thickness = event.at_xpath('thickness'))
+      BigBlueButton.logger.warn("Draw #{draw_id} Shape #{shape[:shape_unique_id]} ID #{shape_id} is missing thickness")
       return
     end
-    if $version_atleast_2_0_0
+    if @version_atleast_2_0_0
       shape[:thickness_percent] = thickness.text.to_f
     else
       shape[:thickness] = thickness.text.to_i
     end
   end
-  if  shape[:type] == 'rectangle' or
-      shape[:type] == 'ellipse' or shape[:type] == 'triangle'
-    fill = event.at_xpath('fill')
-    fill = fill.nil? ? "false" : fill.text
+  if %w[ellipse rectangle triangle].include?(shape_type)
+    fill = event.at_xpath('fill')&.text || 'false'
     shape[:fill] = fill =~ /true/ ? true : false
```

Comment:
Now that I look at it, this code for `shape[:fill]` is broken. It will always return `false`. We don't currently support filled shapes (i'm not sure which BBB versions ever did?), but I guess it never worked. Huh.

Reply:
Shapes can be filled by holding down the control key when running BBB from the develop branch afaik.